### PR TITLE
Add Freesound Audio Tagging 2019

### PR DESCRIPTION
**Pull request by carlthome**
I'd love to see tfds include [the DCASE task that's on Kaggle currently](https://www.kaggle.com/c/freesound-audio-tagging-2019/), but I didn't find any issues or pull requests for adding it so here's my attempt. I'm opening a pull request early so we can pool efforts, in case anyone else would like to make this happen! :koala: 

I still have some reading of the contribution guidelines to do (on it!). Adding `pandas`, and `kaggle` for example, code formatting, unit test framework used by tfds, and so on are on my TODO.

## Loading audio?
`tfds.features.Audio` is really slow (it seems to do single threaded calls to pydub regardless of minibatch size), and I noticed that @adarob's Groove addition just now didn't go via the existing feature connector at all. Is stuff intended to move around? What encoding is preferred for audio datasets? 

Currently this pull request just lets the user consume the file path to the WAV file manually, by returning a `tf.string` with the path to audio file, then a subsequent `tf.io.read_file` and `tf.audio.decode_wav` call can be parallelised in `tf.data.Dataset.map`.

## How to encode multi-label datasets?
Throughout tfds there doesn't seem to be a strong convention for how to encode multi-label datasets, and there doesn't seem to be a feature connector for binary indicator variables yet. I copied the encoding used by the NSynth dataset, where a nested `FeaturesDict` simply contains individual `tf.bool`:s per label, but then it's up to the user to concatenate them which can be bug prone.

## Splitting the dataset
Since the competition is interested in training on noisy labels, and evaluating on curated labels, I think we should default to splitting the dataset in exactly that manner despite the fact that you'd probably rank higher on the Kaggle leaderboard by training on some of the curated data (as that's the same source that's used for the hidden test set).

Another discussion point is how to treat the fact that the test dataset doesn't have any labels. Currently, I'm just returning whatever [0.0,1.0] scores are set in the example_submission.csv provided by the competition.

### CODE REVIEW

File: tensorflow_datasets/audio/freesound_audio_tagging_2019.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
+from pathlib import Path
+
+import kaggle
```
Comment (us) [inline tensorflow_datasets/audio/freesound_audio_tagging_2019.py]: TFDS also have a kaggle downloader on `core/download/kaggle.py`. Please check it.

File: tensorflow_datasets/audio/freesound_audio_tagging_2019.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            tfds.Split.TEST: download_dir / 'sample_submission.csv',
+        }
+
+        zips = {
+            tfds.Split.TRAIN: download_dir / 'train_noisy.zip',
+            tfds.Split.VALIDATION: download_dir / 'train_curated.zip',
+            tfds.Split.TEST: download_dir / 'test.zip',
+        }
+
+        kaggle.api.authenticate()
```
Comment (us) [inline tensorflow_datasets/audio/freesound_audio_tagging_2019.py]: .

File: tensorflow_datasets/audio/freesound_audio_tagging_2019.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
+from pathlib import Path
```
Comment (us) [inline tensorflow_datasets/audio/freesound_audio_tagging_2019.py]: Tensorflow have own io library. Please check this [link](https://www.tensorflow.org/api_docs/python/tf/io).

File: tensorflow_datasets/audio/freesound_audio_tagging_2019.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            urls=_URLS,
+            citation=_CITATION,
+        )
+        return dataset_info
+
+    def _split_generators(self, dl_manager):
+        download_dir = Path(dl_manager._download_dir)
+
+        csvs = {
+            tfds.Split.TRAIN: download_dir / 'train_noisy.csv',
```
Comment (us) [inline tensorflow_datasets/audio/freesound_audio_tagging_2019.py]: I think you miss quotes.

Use `os` library to be compatible with all environments.
Like that `os.path.join('download_dir', 'train_noisy.csv')`

File: tensorflow_datasets/audio/freesound_audio_tagging_2019_test.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        x = tf.transpose(x, (1, 2, 0))
+        spectrogram = tf.abs(x)
+
+        tags = row['tags']
+        tags = tf.stack(list(tags.values()))
+
+        return spectrogram, tags
+
+    dataset = tfds.load(
+        name='freesound_audio_tagging2019',
```
Comment (us) [inline tensorflow_datasets/audio/freesound_audio_tagging_2019_test.py]: You should use fake examples to testing. Check this [link](https://github.com/tensorflow/datasets/blob/master/docs/add_dataset.md#testing-mydataset).

**Comment by us** [comment]
Nice dataset.
Thanks for contributing.

File: tensorflow_datasets/audio/freesound_audio_tagging_2019.py (hunk 3 above)
Comment (Conchylicultor) [reply tensorflow_datasets/audio/freesound_audio_tagging_2019.py]: The Path library is nice but I am worried with Python2 compatibility.

If you want to use it, you would have to add it to the Python2 dependency list: https://github.com/tensorflow/datasets/blob/master/setup.py#L68

File: tensorflow_datasets/audio/freesound_audio_tagging_2019.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
+from pathlib import Path
+
+import kaggle
+import pandas as pd
```
Comment (Conchylicultor) [inline tensorflow_datasets/audio/freesound_audio_tagging_2019.py]: Please have a look at lazy_imports for this: https://www.tensorflow.org/datasets/add_dataset#extra_dependencies

**Comment by Conchylicultor** [comment]
Thank you for your contribution, for the audio encoding, I feel the correct solution would be to do a similar implementation as for `Image`, where the audio file is encoded as string inside the tf record file, and decoded with `tf.audio.decode_wav` inside `tfds.features.Audio.decode_example`.

Like: https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/core/features/image_feature.py#L149

**Comment by us** [comment]
Hey @carlthome, can you check @Conchylicultor pr review

Thanks again!

**Comment by carlthome** [comment]
Closing due to stale
